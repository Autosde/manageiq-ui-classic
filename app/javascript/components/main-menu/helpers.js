export const getHrefByType = (type, href, id) => {
  switch (type) {
    case 'big_iframe':
      return `/dashboard/iframe?id=${id}`;
    case 'modal':
      return undefined;
    default:
      return href;
  }
};

export const getTargetByType = type => (type === 'new_window' ? '_blank' : '_self');

export const handleUnsavedChanges = (e, type) => {
  if (type === 'modal') {
    return sendDataWithRx({ type: 'showAboutModal' });
  }
  return window.miqCheckForChanges();
};
