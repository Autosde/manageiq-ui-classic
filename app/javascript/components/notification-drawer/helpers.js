export const newCountText = function(count) {
  return sprintf(n__('%d New', '%d New'), count);
};

export const getNotficationStatusIconName = notification => (
  {
    info: 'info',
    error: 'error-circle-o',
    danger: 'error-circle-o',
    warning: 'warning-triangle-o',
    success: 'ok',
    ok: 'ok',
  }[notification.type] || '');

export const viewDetails = function(notification) {
  window.location.href = `/restful_redirect?${$.param(notification.data.link)}`;
};
