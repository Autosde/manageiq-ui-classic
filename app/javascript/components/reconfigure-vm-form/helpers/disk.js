import { TYPES, sizeInMB } from './general';

/** Function to execute when the delete backing toggle changed */
export const backupDelete = (id, data, setData) => {
  let backingIds = data.submitParams.diskIds.backup || [];
  const index = id.substring(4);
  if (backingIds.includes(id)) {
    // code to remove from delete backlog
    data.dataTable.disks[index].delete_backing = false;
    backingIds = backingIds.filter((diskId) => diskId !== id);
  } else {
    // code to add the disk to delete backlog
    data.dataTable.disks[index].delete_backing = true;
    backingIds.push(id);
  }
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: data.dataTable.disks,
    },
    submitParams: {
      ...data.submitParams,
      diskIds: {
        ...data.submitParams.diskIds,
        backup: backingIds,
      },
    },
  });
};

/** function to delete disk executed when the delete button gets clicked for the disk datatable */
export const deleteDisk = (row, data, setData) => {
  let dataTableData = data.dataTable.disks;
  let addDisks = data.submitParams.disks.add || [];
  const deleteDisks = data.submitParams.disks.delete || [];
  let addDiskIds = data.submitParams.diskIds.add || [];
  const deleteIds = data.submitParams.diskIds.delete || [];

  if (addDiskIds.includes(row.id)) {
    // delete from the newly added disks
    // eslint-disable-next-line no-alert
    if (window.confirm(__('Are you sure you want to delete this entry?'))) {
      dataTableData = data.dataTable.disks.filter((disk) => disk.id !== row.id);
      addDisks = data.submitParams.disks.add.filter((disk) => disk.id !== row.id);
      addDiskIds = data.submitParams.diskIds.add.filter((diskId) => diskId !== row.id);
      data.changeCounter -= 1;
    }
  } else {
    // remove from the existing disk datatable
    const index = row.id.substring(4);
    const currentRow = data.dataTable.disks[index];
    currentRow.add_remove = 'remove';
    const deleteDiskData = {
      id: row.id,
      disk_name: currentRow.hdFilename,
      delete_backing: deleteIds.includes(row.id),
    };
    deleteDisks.push(deleteDiskData);
    deleteIds.push(row.id);
    data.changeCounter += 1;
  }
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: dataTableData,
    },
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        add: addDisks,
        delete: deleteDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        add: addDiskIds,
        delete: deleteIds,
      },
    },
  });
};

/** Cancel delete disk function
 *  button click from datatable
 */
export const cancelDeleteDisk = (row, data, setData) => {
  const submitDisks = data.submitParams.disks.delete.filter((disk) => disk.id !== row.id);
  const deletedDiskIds = data.submitParams.diskIds.delete.filter((id) => id !== row.id);
  data.changeCounter -= 1;
  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        delete: submitDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        delete: deletedDiskIds,
      },
    },
  });
};

/** Cancel resize disk function - restore the previous data size
 *  button click from datatable
 */
export const cancelResizeDisk = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(4);
  data.dataTable.disks[index].hdSize = data.dataTable.disks[index].old_size;
  data.dataTable.disks[index].hdUnit = data.dataTable.disks[index].old_unit;
  const submitDisks = data.submitParams.disks.resize.filter((disk) => disk.id !== row.id);
  const resizeDiskIds = data.submitParams.diskIds.resize.filter((id) => id !== row.id);
  data.changeCounter -= 1;
  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      disks: {
        ...data.submitParams.disks,
        resize: submitDisks,
      },
      diskIds: {
        ...data.submitParams.diskIds,
        resize: resizeDiskIds,
      },
    },
  });
};

/** Function to check the button need to be disabled or not */
const diskDisableButton = (id, data) => {
  if (data.submitParams.diskIds.delete.includes(id) || data.submitParams.diskIds.add.includes(id)) {
    return true;
  }
  return false;
};

/** Function to check the delete button rendering */
const diskAction = (id, data) => {
  if (data.submitParams.diskIds.delete.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Delete'),
      text: __('Cancel Delete'),
      alt: __('Cancel Delete'),
      kind: 'ghost',
      callback: 'cancelDeleteDisk',
    };
  }
  return {
    is_button: true,
    title: __('Delete'),
    text: __('Delete'),
    alt: __('Delete'),
    kind: 'ghost',
    disabled: data.submitParams.diskIds.resize.includes(id),
    callback: 'deleteDisk',
  };
};

/** Function to check the resize button rendering */
const resizeAction = (id, data) => {
  if (data.submitParams.diskIds.resize.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Resize'),
      text: __('Cancel Resize'),
      alt: __('Cancel Resize'),
      kind: 'ghost',
      callback: 'cancelResizeDisk',
    };
  }
  return {
    is_button: true,
    title: __('Resize'),
    text: __('Resize'),
    alt: __('Resize'),
    kind: 'ghost',
    callback: 'resizeDisk',
    disabled: diskDisableButton(id, data),
  };
};

/** Function to validate the disk size
 * This function will validate the disk size in resize form
 */
export const validateDiskSize = (value, unit, minSize, minUnit) => {
  if (!Number.isNaN(value)) {
    const valueInMb = sizeInMB(value, unit);
    const minSizeInMb = sizeInMB(minSize, minUnit);
    if (valueInMb >= minSizeInMb) {
      return true;
    }
  }
  return false;
};

/** Function to set the Disk datatable row data
 * This formatting is needed for resize function validation
 */
export const formatDiskData = (disks) => {
  disks.forEach((disk) => {
    disk.orgHdSize = disk.hdSize;
    disk.orgHdUnit = disk.hdUnit;
  });

  return disks;
};

/** Function to set the Disk datatable row data  */
export const setDiskData = (disks, roles, data, setData) => {
  const diskData = [];
  disks.forEach((disk, i) => {
    const id = `disk${i}`;
    diskData[i] = {
      id,
      name: disk.hdFilename || ' ',
      type: disk.hdType,
      size: disk.hdSize,
      unit: disk.hdUnit,
      action: diskAction(id, data),
    };
    if (roles.isVmwareInfra) {
      diskData[i].dependent = disk.cb_dependent ? __('Yes') : __('No');
      diskData[i].mode = disk.hdMode || __('None');
      diskData[i].controller = disk.new_controller_type || __('None');
    }
    if (!roles.isVmwareCloud) {
      diskData[i].backing = {
        is_toggle: true,
        labelText: __('Delete Backing'),
        labelA: __('Yes'),
        labelB: __('No'),
        toggled: !!disk.delete_backing,
        ontoggle: () => backupDelete(id, data, setData),
        disabled: diskDisableButton(id, data),
      };
    }
    if (roles.isRedhat) {
      diskData[i].bootable = disk.cb_bootable === true ? __('Yes') : __('No');
    }
    if (roles.allowDiskSizeChange) {
      diskData[i].resize = resizeAction(id, data);
    }
  });
  return diskData;
};

/** Function to set the Disk datatable headers  */
export const getDiskTableHeaders = (roles) => {
  const headers = [
    { key: 'name', header: __('Name') },
    { key: 'type', header: __('Type') },
    { key: 'size', header: __('Size') },
    { key: 'unit', header: __('Unit') },
  ];
  if (roles.isVmwareInfra) {
    headers.push(
      { key: 'mode', header: __('Mode') },
      { key: 'controller', header: __('Controller Type') },
      { key: 'dependent', header: __('Dependent') }
    );
  }
  if (!roles.isVmwareCloud) {
    headers.push({ key: 'backing', header: __('Delete Backing') });
  }
  if (roles.isRedhat) {
    headers.push({ key: 'bootable', header: __('Bootable') });
  }
  if (roles.allowDiskSizeChange) {
    headers.push({ key: 'resize', header: __('Resize') });
  }
  headers.push({ key: 'action', header: __('Action'), contentButton: true });
  return headers;
};

/** function to execute disk sub forms submit gets called
 * Disk add, resize function
*/
export const setDiskFormSubmit = (data, setData, formData, roles, renderData) => {
  const dataTableData = data.dataTable.disks || [];
  if (data.form.action === 'add') {
    // code for add
    const id = `disk${dataTableData.length}`;
    const diskDatatableData = {
      id,
      hdFilename: '',
      hdType: formData.type,
      hdSize: formData.size,
      hdUnit: formData.unit,
      add_remove: data.form.action,
    };
    const submitDiskData = {
      id,
      disk_size_in_mb: sizeInMB(formData.size, formData.unit),
      thin_provisioned: formData.type === 'thin',
      type: formData.type,
    };
    if (roles.isVmwareInfra) {
      diskDatatableData.hdMode = formData.mode;
      diskDatatableData.new_controller_type = formData.controller;
      diskDatatableData.cb_dependent = formData.dependent || false;
      submitDiskData.persistent = formData.hdMode;
      submitDiskData.new_controller_type = formData.controller;
      submitDiskData.dependent = formData.dependent || false;
    }
    if (!roles.isVmwareCloud) {
      diskDatatableData.backing = false;
    }
    if (roles.isRedhat) {
      diskDatatableData.cb_bootable = formData.bootable || false;
      submitDiskData.bootable = formData.bootable || false;
    }
    data.submitParams.disks.add.push(submitDiskData);
    data.submitParams.diskIds.add.push(diskDatatableData.id);
    dataTableData.push(diskDatatableData);
  } else {
    // code for resize option
    const { id } = data.editingRow;
    const index = id.substring(4);
    dataTableData[index].old_size = dataTableData[index].hdSize;
    dataTableData[index].hdSize = formData.size;
    dataTableData[index].old_unit = dataTableData[index].hdUnit;
    dataTableData[index].hdUnit = formData.unit;
    const resizeData = {
      id,
      disk_name: dataTableData[index].hdFilename,
      disk_size_in_mb: sizeInMB(formData.size, formData.unit),
    };
    data.submitParams.disks.resize.push(resizeData);
    data.submitParams.diskIds.resize.push(id);
  }
  data.changeCounter += 1;
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      disks: dataTableData,
    },
    form: {
      title: renderData.title,
      type: TYPES.RECONFIGURE,
      className: renderData.className,
    },
  });
};
