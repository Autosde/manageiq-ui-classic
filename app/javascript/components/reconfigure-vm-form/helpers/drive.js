import { TYPES } from './general';

/** Function to cancel the connection of the drive from device  */
export const cancelConnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  datatableData[index].filename = datatableData[index].old_filename;

  const submitDrives = data.submitParams.drives.connect.filter((drive) => drive.id !== id);
  const connectedDriveIds = data.submitParams.driveIds.connect.filter((driveId) => driveId !== id);
  data.changeCounter -= 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        connect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        connect: connectedDriveIds,
      },
    },
  });
};

/** Function to cancel the disconnection of drives from device  */
export const cancelDisconnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  datatableData[index].filename = datatableData[index].disconnect_filename;

  const submitDrives = data.submitParams.drives.disconnect.filter((drive) => drive.id !== id);
  const disConnectedDriveIds = data.submitParams.driveIds.disconnect.filter((driveId) => driveId !== id);
  data.changeCounter -= 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        disconnect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        disconnect: disConnectedDriveIds,
      },
    },
  });
};

/** Function to disconnect drives from device  */
export const disconnectDrives = (row, data, setData) => {
  const { id } = row;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  const submitDrives = data.submitParams.drives.disconnect || [];
  const disConnectedDriveIds = data.submitParams.driveIds.disconnect || [];
  datatableData[index].disconnect_filename = datatableData[index].filename;
  datatableData[index].filename = '';
  const disconnectDriveData = {
    id,
    device_name: datatableData[index].device_name,
  };
  submitDrives.push(disconnectDriveData);
  disConnectedDriveIds.push(id);
  data.changeCounter += 1;
  setData({
    ...data,
    isLoading: false,
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        disconnect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        disconnect: disConnectedDriveIds,
      },
    },
  });
};

/** Function to check the connect button render options  */
const connectAction = (id, file, data) => {
  if (data.submitParams.driveIds.connect.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Connect'),
      text: __('Cancel Connect'),
      alt: __('Cancel Connect'),
      kind: 'ghost',
      callback: 'cancelConnectDrives',
    };
  }
  return {
    is_button: true,
    title: __('Connect'),
    text: __('Connect'),
    alt: __('Connect'),
    disabled: !!file,
    kind: 'ghost',
    callback: 'connectDrives',
  };
};

/** Function to check the disconnect button render options  */
const disConnectAction = (id, data) => {
  if (data.submitParams.driveIds.disconnect.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Disconnect'),
      text: __('Cancel Disconnect'),
      alt: __('Cancel Disconnect'),
      kind: 'ghost',
      callback: 'cancelDisconnectDrives',
      disabled: data.submitParams.driveIds.connect.includes(id),
    };
  }
  return {
    is_button: true,
    title: __('Disconnect'),
    text: __('Disconnect'),
    alt: __('Disconnect'),
    kind: 'ghost',
    callback: 'disconnectDrives',
    disabled: data.submitParams.driveIds.connect.includes(id),
  };
};

/** Function to set the CD/ROM Drive datatable row data  */
export const setDrivesData = (drives, data) => {
  const driveData = [];

  drives.forEach((drive, i) => {
    const id = `drive${i}`;
    driveData[i] = {
      id,
      name: drive.device_name,
      hostFile: drive.filename,
      action: connectAction(id, drive.filename, data),
    };
    if (drive.filename || drive.disconnect_filename) {
      driveData[i].disconnect = disConnectAction(id, data);
    } else {
      driveData[i].disconnect = '';
    }
  });

  return driveData;
};

/** Function to set the CD-ROM drive datatable headers  */
export const getDriveTableHeaders = () => [
  { key: 'name', header: __('Name') },
  { key: 'hostFile', header: __('Host File') },
  { key: 'disconnect', header: __('Disconnect') },
  { key: 'action', header: __('Actions') },
];

/** function to execute drive sub forms submit gets called
 * Drive connect form submission
*/
export const setDriveFormSubmit = (data, setData, formData, renderData) => {
  const { id } = data.editingRow;
  const index = id.substring(5);
  const datatableData = data.dataTable.drives;
  const submitDrives = data.submitParams.drives.connect || [];
  const connectedDriveIds = data.submitParams.driveIds.connect || [];
  const isoSelect = formData.host_file.split(',');

  datatableData[index].old_filename = datatableData[index].filename;
  // eslint-disable-next-line prefer-destructuring
  datatableData[index].filename = isoSelect[0];

  const connectDriveData = {
    id,
    device_name: datatableData[index].device_name,
    filename: datatableData[index].filename,
    storage_id: isoSelect[1],
  };
  submitDrives.push(connectDriveData);
  connectedDriveIds.push(id);
  data.changeCounter += 1;

  setData({
    ...data,
    isLoading: false,
    form: {
      title: renderData.title,
      type: TYPES.RECONFIGURE,
      className: renderData.className,
    },
    dataTable: {
      ...data.dataTable,
      drives: datatableData,
    },
    submitParams: {
      ...data.submitParams,
      drives: {
        ...data.submitParams.drives,
        connect: submitDrives,
      },
      driveIds: {
        ...data.submitParams.driveIds,
        connect: connectedDriveIds,
      },
    },
  });
};
