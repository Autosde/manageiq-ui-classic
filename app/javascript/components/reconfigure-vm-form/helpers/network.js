import { TYPES, formsData } from './general';

/** Function which executes when the cancel edit button gets clicked */
export const cancelEditNetwork = (row, data, setData, roles) => {
  const index = row.id.substring(7);
  const networkRows = data.dataTable.networkAdapters;

  if (!roles.isVmwareCloud) {
    networkRows[index].vlan = networkRows[index].old_vlan;
  }

  if (roles.isVmwareCloud) {
    networkRows[index].network = networkRows[index].old_network;
  }
  const submitNetworks = data.submitParams.networks.edit.filter((network) => network.id !== row.id);
  const editedNetworkIds = data.submitParams.networkIds.edit.filter((id) => id !== row.id);

  data.dataTable.networkAdapters = networkRows;
  data.changeCounter -= 1;

  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      networks: {
        ...data.submitParams.networks,
        edit: submitNetworks,
      },
      networkIds: {
        ...data.submitParams.networkIds,
        edit: editedNetworkIds,
      },
    },
  });
};

/** Function which executes when the delete button gets clicked */
export const deleteNetwork = (row, data, setData, roles) => {
  let dataTableData = data.dataTable.networkAdapters || [];
  let submitNetworks = data.submitParams.networks.add || [];
  let addNetworkIds = data.submitParams.networkIds.add || [];
  const deleteNetworksData = data.submitParams.networks.delete || [];
  const deletedNetworkIds = data.submitParams.networkIds.delete || [];

  if (data.submitParams.networkIds.add.includes(row.id)) {
    // remove from the newly added networks
    // eslint-disable-next-line no-alert
    if (window.confirm(__('Are you sure you want to delete this entry?'))) {
      dataTableData = data.dataTable.networkAdapters.filter((network) => network.id !== row.id);
      submitNetworks = data.submitParams.networks.add.filter((network) => network.id !== row.id);
      addNetworkIds = data.submitParams.networkIds.add.filter((id) => id !== row.id);
    }
    data.changeCounter -= 1;
  } else {
    // remove from the existing networks
    const index = row.id.substring(7);
    const deleteNetworkData = {
      id: row.id,
      network: {
        name: dataTableData[index].name,
        mac: dataTableData[index].mac,
        add_remove: 'remove',
      },
    };
    if (!roles.isVmwareCloud) {
      deleteNetworkData.network.vlan = dataTableData[index].vlan;
    }
    if (roles.isVmwareCloud) {
      deleteNetworkData.network.network = dataTableData[index].network;
    }
    deleteNetworksData.push(deleteNetworkData);
    deletedNetworkIds.push(row.id);
    data.changeCounter += 1;
  }
  setData({
    ...data,
    dataTable: {
      ...data.dataTable,
      networkAdapters: dataTableData,
    },
    submitParams: {
      ...data.submitParams,
      networks: {
        ...data.submitParams.networks,
        add: submitNetworks,
        delete: deleteNetworksData,
      },
      networkIds: {
        ...data.submitParams.networkIds,
        add: addNetworkIds,
        delete: deletedNetworkIds,
      },
    },
  });
};

/** Function which executes when the cancel delete button gets clicked */
export const cancelDeleteNetwork = (row, data, setData) => {
  const deleteNetworksData = data.submitParams.networks.delete.filter((network) => network.id !== row.id);
  const deletedNetworkIds = data.submitParams.networkIds.delete.filter((id) => id !== row.id);
  data.changeCounter -= 1;
  setData({
    ...data,
    submitParams: {
      ...data.submitParams,
      networks: {
        ...data.submitParams.networks,
        delete: deleteNetworksData,
      },
      networkIds: {
        ...data.submitParams.networkIds,
        delete: deletedNetworkIds,
      },
    },
  });
};

/** Function to set the Network datatable row data - name */
const getNetworkName = (id, data) => {
  if (data.submitParams.networkIds.add.includes(id)) {
    return __('to be determined');
  }
  return __(' ');
};

/** Function to set the Network datatable row data - mac */
const getNetworkMac = (id, data) => {
  if (data.submitParams.networkIds.add.includes(id)) {
    return __('not available yet');
  }
  return __(' ');
};

/** Function to check edit button should be disabled or not */
const disableNetworkEditButton = (id, data) => data.submitParams.networkIds.add.includes(id) || data.submitParams.networkIds.delete.includes(id);

/** Function to check and render the network edit button */
const networkEditAction = (id, data) => {
  if (data.submitParams.networkIds.edit.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Edit'),
      text: __('Cancel Edit'),
      alt: __('Cancel Edit'),
      kind: 'ghost',
      callback: 'cancelEditNetwork',
      disabled: disableNetworkEditButton(id, data),
    };
  }
  return {
    is_button: true,
    title: __('Edit'),
    text: __('Edit'),
    alt: __('Edit'),
    kind: 'ghost',
    disabled: disableNetworkEditButton(id, data),
    callback: 'editNetwork',
  };
};

/** Function to check and render the network delete button */
const networkDeleteAction = (id, data) => {
  if (data.submitParams.networkIds.delete.includes(id)) {
    return {
      is_button: true,
      title: __('Cancel Delete'),
      text: __('Cancel Delete'),
      alt: __('Cancel Delete'),
      kind: 'ghost',
      callback: 'cancelDeleteNetwork',
    };
  }
  return {
    is_button: true,
    title: __('Delete'),
    text: __('Delete'),
    alt: __('Delete'),
    kind: 'ghost',
    callback: 'deleteNetwork',
    disabled: data.submitParams.networkIds.edit.includes(id),
  };
};

/** Function to set the Network datatable row data  */
export const setNetworkData = (networks, roles, data) => {
  const networkData = [];
  networks.forEach((network, i) => {
    const id = `network${i}`;
    const name = network.name ? network.name : getNetworkName(id, data);
    const mac = network.mac ? network.mac : getNetworkMac(id, data);
    networkData[i] = {
      id,
      name,
      mac,
      action: networkDeleteAction(id, data),
      edit: networkEditAction(id, data),
    };
    if (!roles.isVmwareCloud) {
      networkData[i].vlan = network.vlan;
    }
    if (roles.isVmwareCloud) {
      networkData[i].network = network.network || __('None');
    }
  });
  return networkData;
};

/** Function to set the Network datatable headers  */
export const getNetworkTableHeaders = (roles) => {
  const headers = [
    { key: 'name', header: __('Name') },
    { key: 'mac', header: __('MAC address') },
  ];
  if (!roles.isVmwareCloud) {
    headers.push({ key: 'vlan', header: __('vLan') });
  }
  if (roles.isVmwareCloud) {
    headers.push({ key: 'network', header: __('Network') });
  }
  headers.push({ key: 'edit', header: __('Edit'), contentButton: true }, { key: 'action', header: __('Action'), contentButton: true });
  return headers;
};

/** Setting up the network form details when the form loads for cloud */
export const setNetworkOptions = (id, data, setData, type, action, row) => {
  API.get(`/api/cloud_networks?expand=resources&attributes=name&filter[]=orchestration_stack_id=${id}`).then(({ resources }) => {
    const networkOptions = resources.map((item) => (
      item.name
    ));
    setData({
      ...data,
      isLoading: false,
      networkOptions,
      editingRow: row,
      form: {
        title: formsData[type].title,
        type,
        action,
        className: formsData[type].className,
      },
    });
  });
};

/** Function to execute the network form data submission
 * network add, edit function
*/
export const setNetworkFormSubmit = (data, setData, formData, roles, renderData) => {
  if (data.form.action === 'add') {
    // code for add function
    const id = `network${data.dataTable.networkAdapters.length}`;
    const networkData = {
      id,
      name: '',
    };
    const networkSubmitData = {
      id,
      name: 'to be determined',
    };
    if (!roles.isVmwareCloud) {
      networkData.vlan = formData.vlan;
      networkSubmitData.network = formData.vlan;
    }
    if (roles.isVmwareCloud) {
      networkData.network = formData.network;
      networkSubmitData.cloud_network = formData.network;
    }
    data.submitParams.networks.add.push(networkSubmitData);
    data.submitParams.networkIds.add.push(id);
    data.dataTable.networkAdapters.push(networkData);
  } else {
    // code for edit function
    const index = data.editingRow.id.substring(7);
    const networkData = { id: data.editingRow.id };
    const networkRows = data.dataTable.networkAdapters;
    if (!roles.isVmwareCloud) {
      networkData.network = formData.vlan;
      networkRows[index].old_vlan = networkRows[index].vlan;
      networkRows[index].vlan = formData.vlan;
    }
    if (roles.isVmwareCloud) {
      networkData.network = formData.network;
      networkRows[index].old_network = networkRows[index].network;
      networkRows[index].network = networkData.network;
    }
    networkData.name = networkRows[index].name;
    data.submitParams.networks.edit.push(networkData);
    data.submitParams.networkIds.edit.push(networkData.id);
    data.dataTable.networkAdapters = networkRows;
  }
  data.changeCounter += 1;
  setData({
    ...data,
    isLoading: false,
    form: {
      title: renderData.title,
      type: TYPES.RECONFIGURE,
      className: renderData.className,
    },
  });
};
