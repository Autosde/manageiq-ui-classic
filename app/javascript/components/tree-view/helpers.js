import { Tree } from 'react-wooden-tree';

/**
 * Helper
 * Changes class to className.
 * Sets undefined checkbox to null.
 * Moves the key to attr.key.
 */
const convert = (tree, fn) => tree.map((n) => {
  let node = { ...n };
  if (node.class) {
    node = { ...node, classes: node.class };
    delete node.class;
  }

  // The server cannot send null, just undefined, this is why it is tested as a string.
  if (node.state.checked === 'undefined') {
    node = { ...node, state: { ...node.state, checked: null } };
  } else if (node.state.checked === null) {
    node = { ...node, state: { ...node.state, checked: false } };
  }

  if (node.key) {
    node = { ...node, attr: { key: node.key } };
    delete node.key;
  }

  if (node.nodes) {
    node = { ...node, nodes: convert(node.nodes, fn) };
  }

  return fn ? fn(node) : node;
});

const flatten = data => Tree.convertHierarchicalTree(Tree.initHierarchicalTree(data));

/**
 * Helper
 * Activates the first active node in the tree, which is passed trought the props.
 *
 * FIXME Delete this function when bakcend is sending data where the initial node is activated,
 * or when all the trees calling a reducer if they want to have an activated node at the start.
 */

const activateNode = (tree, doActivate, key) => {
  if (!doActivate) {
    return tree;
  }

  let node = Tree.nodeSelector(tree, Tree.nodeSearch(tree, null, 'key', key)[0]);
  node = { ...node, state: { ...node.state, selected: true } };
  return { ...tree, [node.nodeId]: node };
};

const callBack = (nodeId, type, value, namespace) => ({ nodeId, type, value, namespace });

export { convert, flatten, activateNode, callBack };
