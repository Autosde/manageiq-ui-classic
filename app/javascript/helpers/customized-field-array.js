import React from 'react';
import { FieldArray } from '@@ddf';
import FormRenderer, { useFieldApi, useFormApi } from '@data-driven-forms/react-form-renderer';
import mapper from '../forms/mappers/componentMapper';

const ArrayItem = ({ fields, name }) => {
  const { renderForm } = useFormApi();
  const editedFields = fields.map((field) => ({ ...field, name: `${name}.${field.name}` }));

  return (
    <>
      {renderForm(editedFields)}
      <br />
    </>
  );
};

const FieldArrayCustom = (props) => {
  const {
    fieldKey, fields, itemDefault, ...rest
  } = useFieldApi(props);

  return (
    <FieldArray key={fieldKey} name={rest.input.name}>
      {(array) => (
        <>
          {array.fields.map((name, index) => (
            <ArrayItem
              key={`${name || fieldKey}`}
              fields={fields}
              name={name}
              fieldKey={fieldKey}
              fieldIndex={index}
            />
          ))}
          {(array.fields.length === 0) && array.fields.push(itemDefault)}
        </>
      )}
    </FieldArray>
  );
};

const componentMapper = {
  ...mapper,
  'custom-field-array': FieldArrayCustom,
};

const CustomizedFieldArray = async(arrayName, arrayFields) => {
  const getFields = (arrayFields) => arrayFields.map((field) => ({
    component: field.component,
    id: field.id,
    name: field.name,
    label: field.label,
    options: field.options,
  }));

  const schema = {
    fields: [
      {
        component: 'custom-field-array',
        name: arrayName,
        fieldKey: 'field_array',
        title: arrayName,
        fields: getFields(arrayFields),
      },
    ],
  };

  return (
    <>
      <FormRenderer
        componentMapper={componentMapper}
        schema={schema}
      />
    </>
  );
};

export { CustomizedFieldArray };
