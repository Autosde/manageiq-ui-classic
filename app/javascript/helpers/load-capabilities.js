import Api from '../http_api/api';

// in providers capabilities jsonb field, capabilities are saved like {'abstract_capability': 'name', 'uuid': '999..', 'value': 'true/false'},
// but in the storage_manager models the capability name is saved under different keys, instead of 'abstract_capability'.
// the parameter fieldName enables to dynamically access the correct key.

const loadCapabilities = (sourceType, sourceId, fieldName) => Api.get(`/api/${sourceType}/${sourceId}?attributes=capabilities`)
  .then(({ capabilities }) => capabilities.map((capability) => (
    { value: capability.uuid, label: `${capability[`${fieldName}`]}: ${capability.value}` })));

export default loadCapabilities;
