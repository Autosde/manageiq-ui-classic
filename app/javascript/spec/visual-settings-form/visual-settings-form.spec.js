import React from 'react';
import toJson from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import VisualSettingsForm from '../../components/visual-settings-form';
import { mount } from '../helpers/mountForm';

describe('visual settings form', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  window.locales = [];

  const shortcuts = '/api/shortcuts?expand=resources&attributes=description,url';
  const users = '/api/users/1?attributes=settings';

  it('calls the API endpoints to preseed the form', async(done) => {
    fetchMock.get(shortcuts, {
      resources: [
        {
          url: 'foo',
          description: 'bar',
        },
      ],
    });

    fetchMock.get(users, { settings: {} });

    fetchMock.get('/api', { timezones: [] });

    let wrapper;
    await act(async() => {
      wrapper = mount(<VisualSettingsForm recordId="1" />);
    });

    wrapper.update();

    expect(fetchMock.called(shortcuts)).toBe(true);
    expect(fetchMock.called(users)).toBe(true);
    expect(fetchMock.called('/api')).toBe(true);

    done();
  });

  it('matches the snapshot', async(done) => {
    let wrapper;
    await act(async() => {
      wrapper = shallow(<VisualSettingsForm recordId="1" />);
    });

    expect(toJson(wrapper)).toMatchSnapshot();
    done();
  });
});
