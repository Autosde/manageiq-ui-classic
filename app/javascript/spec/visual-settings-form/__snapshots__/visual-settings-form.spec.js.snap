// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visual settings form matches the snapshot 1`] = `""`;
